File: Assignment 4/task1.scala
import org.apache.spark.sql.SparkSession
import org.graphframes.GraphFrame

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

/**
 * Task1.scala
 *
 * DSCI-553: Assignment 4 (Spring 2025)
 * -------------------------------------
 * Goal: Use GraphFrames + Label Propagation (LPA)
 *       to detect communities in an undirected graph,
 *       based on users who share >= threshold businesses.
 *
 * 1) We read input CSV with (user_id, business_id).
 * 2) Build user -> set(business) map.
 * 3) For each pair of users, if overlap >= threshold, we add an edge.
 * 4) Create a GraphFrame with these edges (undirected) & nodes.
 * 5) Run LPA with maxIter=5 to detect communities.
 * 6) Sort communities by (size asc, then lexicographically by smallest user).
 * 7) Write each community line => 'user1', 'user2', 'user3', ...
 *
 * This Scala code mirrors the working Python logic, ensuring
 * only users who share at least one threshold-based edge are included.
 *
 * Example Usage (spark-submit):
 *   spark-submit \
 *     --class Task1 \
 *     target/scala-2.12/hw4.jar \
 *     7 data/ub_sample_data.csv output_task1.txt
 */
object task1 {
  def main(args: Array[String]): Unit = {

    // -----------------------------------------------------------------------
    //  (A) Parse Command-Line Arguments
    // -----------------------------------------------------------------------
    // The assignment expects:
    //  arg(0) = filterThreshold (int)
    //  arg(1) = inputFilePath (string)
    //  arg(2) = outputFilePath (string)
    if (args.length != 3) {
      println("Usage: Task1 <filter_threshold> <input_file_path> <output_file_path>")
      sys.exit(1)
    }

    // Convert the first arg to an integer threshold
    val filterThreshold = args(0).toInt
    // The CSV file path
    val inputPath       = args(1)
    // The final output text file path
    val outputPath      = args(2)

    // -----------------------------------------------------------------------
    //  (B) Initialize SparkSession
    // -----------------------------------------------------------------------
    // "local[*]" means we use all available cores on this machine,
    // but you can omit master() and let spark-submit specify it.
    val spark = SparkSession.builder()
      .appName("Task1-GraphFrames")
      .getOrCreate()

    // We'll reduce Spark's console logs to "ERROR" only:
    val sc = spark.sparkContext
    sc.setLogLevel("ERROR")

    // -----------------------------------------------------------------------
    //  (C) Read & Preprocess the Input CSV
    // -----------------------------------------------------------------------
    // The CSV has a header, so we specify .option("header", "true").
    // Then we select the "user_id" and "business_id" columns to remain consistent.
    val rawDF = spark.read
      .option("header", "true")
      .csv(inputPath)

    // We convert each row => (String user, String business).
    // This is an RDD of (user_id, business_id) pairs.
    val userBizRDD = rawDF.rdd.map { row =>
      val user = row.getString(0)   // user_id
      val biz  = row.getString(1)   // business_id
      (user, biz)
    }

    // -----------------------------------------------------------------------
    //  (D) Build a Map of user -> Set[businesses]
    // -----------------------------------------------------------------------
    // We'll group by user and collect the businesses as a set, which
    // helps us quickly do an intersection to check overlap.
    // Then .collectAsMap() pulls it to the driver.
    val userBizMap = userBizRDD
      .groupByKey()
      .mapValues(_.toSet)
      .collectAsMap()

    // -----------------------------------------------------------------------
    //  (E) Gather All Potential Users from userBizMap
    // -----------------------------------------------------------------------
    // In Python, we used the same approach to avoid including users
    // that have NO edges. userBizMap.keySet simply has all users
    // that appear in the CSV. We'll filter down further when we build edges.
    val userArray = userBizMap.keySet.toArray.sorted

    // -----------------------------------------------------------------------
    //  (F) Identify Undirected Edges Based on Overlap
    // -----------------------------------------------------------------------
    // If two users share >= filterThreshold businesses,
    // we add an edge in both directions to edgesList
    // and also add them to nodeSet.
    val edgesList = new ArrayBuffer[(String, String)]()
    val nodeSet   = mutable.Set[String]()

    for (i <- userArray.indices; j <- i + 1 until userArray.length) {
      val u1 = userArray(i)
      val u2 = userArray(j)

      // Retrieve each user's set of businesses
      val busA = userBizMap(u1)
      val busB = userBizMap(u2)

      // Intersection
      val commonCount = busA.intersect(busB).size

      // If the overlap is >= threshold, there's an edge
      if (commonCount >= filterThreshold) {
        // Add edges in both directions (u1->u2, u2->u1)
        edgesList += ((u1, u2))
        edgesList += ((u2, u1))

        // Mark them as relevant nodes
        nodeSet += u1
        nodeSet += u2
      }
    }

    // -----------------------------------------------------------------------
    //  (G) Create GraphFrame (nodes, edges)
    // -----------------------------------------------------------------------
    // We'll convert nodeSet => a DataFrame column "id"
    // and edgesList => a DataFrame with columns "src" and "dst".
    import spark.implicits._
    val nodesDF = nodeSet.toSeq.toDF("id")
    val edgesDF = edgesList.toSeq.toDF("src", "dst")

    // Build the GraphFrame, which can be used to run LPA.
    val gf = GraphFrame(nodesDF, edgesDF)

    // -----------------------------------------------------------------------
    //  (H) Run Label Propagation Algorithm (LPA) with maxIter=5
    // -----------------------------------------------------------------------
    // This returns a DataFrame with columns ["id", "label"].
    // "label" is a long representing the community ID after LPA.
    val lpaDF = gf.labelPropagation.maxIter(5).run()

    // -----------------------------------------------------------------------
    //  (I) Group the LPA results => communities
    // -----------------------------------------------------------------------
    // We do a standard approach:
    //   - RDD of (label, user)
    //   - groupByKey => label => multiple users
    //   - sort each community's user IDs
    //   - drop the label, keep just the sorted user-list
    //   - finally, sort the entire set of communities by:
    //        1) ascending length
    //        2) lexicographically smallest user
    val communitiesRDD = lpaDF.rdd
      .map { row =>
        val user = row.getAs[String]("id")
        val labl = row.getAs[Long]("label")
        (labl, user)
      }
      .groupByKey()
      .mapValues(users => users.toList.sorted)
      .map(_._2) // drop the label, keep the sorted user list
      .sortBy(community => (community.size, community.headOption.getOrElse("")))

    // -----------------------------------------------------------------------
    //  (J) Write Communities to Output File
    // -----------------------------------------------------------------------
    // Each line => 'userA', 'userB', ...
    // We'll create a PrintWriter => output file, then print each community line.
    val finalComms = communitiesRDD.collect()
    val pw = new java.io.PrintWriter(outputPath)

    for (comm <- finalComms) {
      // We wrap each user in single quotes, joined by ", "
      val line = comm.map(u => s"'$u'").mkString(", ")
      pw.println(line)
    }

    pw.close()

    // -----------------------------------------------------------------------
    //  (K) Cleanup
    // -----------------------------------------------------------------------
    // Stop the Spark session. We’re done!
    spark.stop()
  }
}

File: Assignment 4/task1.py
"""
=================================================================================
    DSCI-553: Assignment 4 (Spring 2025)
    Task 1: Community Detection using GraphFrames & Label Propagation Algorithm
---------------------------------------------------------------------------------
    OVERVIEW:
    1) We read the CSV input containing user_id,business_id (with a header).
    2) For each user, gather the set of businesses they reviewed.
    3) Construct an undirected graph where:
        - Each distinct user is a node.
        - An edge between two users exists only if they have >= filter_threshold
          common businesses reviewed.
    4) Use GraphFrames + labelPropagation(maxIter=5) to detect communities.
    5) Sort the discovered communities by:
        (a) The ascending size of each community (smaller sets first).
        (b) The lexicographically smallest user in that community.
        Then, within a community, sort the user IDs in ascending order.
    6) Output each community on a new line. Each line's format:
        'userA', 'userB', 'userC', ...
---------------------------------------------------------------------------------
    EXECUTION EXAMPLE:
        spark-submit --packages graphframes:graphframes:0.8.2-spark3.1-s_2.12 \
                     task1.py 7 /path/to/ub_sample_data.csv /path/to/output.txt
---------------------------------------------------------------------------------
    NOTE:
     - We rely on GraphFrames 0.8.2 for Spark 3.1.2.
     - This code is purely for Task 1. Do NOT use it for Task 2, as Task 2
       requires strictly Spark RDD (no DataFrames, no GraphFrames).
=================================================================================
"""

import sys
import time
import os
from itertools import combinations

# -------------- SPARK IMPORTS -------------- #
from pyspark import SparkContext
from pyspark.sql import SparkSession, SQLContext
from graphframes import GraphFrame
# ------------------------------------------- #
# ------------------------------------------------------------------------------
# We specify the GraphFrames package, so Spark knows where to fetch it. 
# (You can also provide this via 'spark-submit --packages ...')
# ------------------------------------------------------------------------------
os.environ["PYSPARK_SUBMIT_ARGS"] = "--packages graphframes:graphframes:0.8.2-spark3.1-s_2.12 pyspark-shell"

def main():
    """
    Main entry point for Task 1.
    1) Parse command-line args
    2) Read data, build user-business mapping
    3) Construct graph using threshold
    4) Run LPA to detect communities
    5) Output final results
    6) Print duration
    """
    # -------------------------
    # 1. Parse command-line arguments
    # -------------------------
    # Expecting 3 arguments in total:
    #   1) filter_threshold  (integer)
    #   2) input_file_path   (str)
    #   3) output_file_path  (str)
    # Example: spark-submit ... task1.py 7 input.csv output.txt
    if len(sys.argv) != 4:
        print("Usage: task1.py <filter_threshold> <input_file_path> <output_file_path>")
        sys.exit(-1)

    filter_threshold = int(sys.argv[1])
    input_file_path  = sys.argv[2]
    output_file_path = sys.argv[3]

    # Start measuring runtime
    start_time = time.time()

    # -------------------------------------
    # 2. Initialize Spark Context & Session
    # -------------------------------------
    # We create a SparkContext for RDD operations,
    # and a SparkSession (or SQLContext) for DataFrame operations.
    sc = SparkContext(appName="task1")
    sc.setLogLevel("ERROR")  # reduce Spark logging in console
    spark = SparkSession.builder \
        .config("spark.some.config.option", "some-value") \
        .getOrCreate()
    sqlContext = SQLContext(sc)

    # -------------------------------------
    # 3. Load and preprocess the input CSV
    # -------------------------------------
    # We'll read lines from the CSV, skip the header, then parse each row into (user_id, business_id).
    # Example of row: "user_id,business_id"
    # We do an initial textFile read, remove the header, then split by comma.
    input_rdd = sc.textFile(input_file_path)
    header = input_rdd.first()
    data_rdd = input_rdd.filter(lambda line: line != header)\
                        .map(lambda line: line.strip().split(","))  # => (user_id, business_id)

    # -------------------------------------
    # 4. Build dictionary: user -> set of businesses
    # -------------------------------------
    # We want something like: user_business_map[user] = {business1, business2, ...}
    # Steps:
    #   - map each row -> (user_id, business_id)
    #   - groupByKey => user_id => list of business_ids
    #   - convert the list to a set for easy intersection
    #
    # We collectAsMap() to bring it to the driver. This can be big, but we’ll assume
    # dataset is small enough for assignment. In a real system, we might do a different approach.
    user_business_map = data_rdd \
        .map(lambda x: (x[0], x[1])) \
        .groupByKey() \
        .mapValues(lambda biz_list: set(biz_list)) \
        .collectAsMap()

    # We'll also get a distinct list of all users.
    distinct_users = data_rdd.map(lambda x: x[0]).distinct().collect()

    # -------------------------------------
    # 5. Identify all edges based on threshold
    # -------------------------------------
    # We want an edge between userA and userB if they share at least
    # 'filter_threshold' number of businesses. 
    # We'll do a combination of all distinct users two at a time
    # because the PDF says each edge is undirected, unweighted.
    #
    #    for pair in combinations(distinct_users, 2):
    #       if overlap >= filter_threshold => create an edge in both directions
    #
    # We'll store edges in a list of tuples (src, dst).
    # Then we will create a DataFrame from that list for the GraphFrame.
    from itertools import combinations
    edges_list = []
    # We'll store a set of nodes that appear, so we only keep relevant users in final graph
    nodes_set = set()

    for userA, userB in combinations(distinct_users, 2):
        # Compute intersection size
        businessesA = user_business_map[userA]
        businessesB = user_business_map[userB]
        common_count = len(businessesA.intersection(businessesB))

        if common_count >= filter_threshold:
            # Add these two users to our nodes set
            nodes_set.add(userA)
            nodes_set.add(userB)
            # Because GraphFrames require directed edges for construction,
            # we add both (userA -> userB) and (userB -> userA).
            edges_list.append((userA, userB))
            edges_list.append((userB, userA))

    # Now convert nodes_set into a list so we can build a DataFrame.
    nodes_list = list(nodes_set)

    # -------------------------------------
    # 6. Create DataFrames for GraphFrames
    # -------------------------------------
    # GraphFrame requires:
    #   - A vertices DataFrame with a column "id" 
    #   - An edges DataFrame with columns "src" and "dst"
    #
    # We'll create them from the Python lists:
    #    nodes_list -> Spark DataFrame ["id"]
    #    edges_list -> Spark DataFrame ["src", "dst"]
    nodes_df = sqlContext.createDataFrame([(n,) for n in nodes_list], ["id"])
    edges_df = sqlContext.createDataFrame(edges_list, ["src", "dst"])

    # -------------------------------------
    # 7. Construct the GraphFrame and run LPA
    # -------------------------------------
    # Label Propagation Algorithm:
    #   - We use maxIter=5 (as required by the PDF).
    #   - The result is a DataFrame of columns: ["id", "label"]
    #     The "label" indicates the community ID.
    GF = GraphFrame(nodes_df, edges_df)
    lpa_result_df = GF.labelPropagation(maxIter=5)

    # -------------------------------------
    # 8. Group the LPA results into communities
    # -------------------------------------
    # The LPA result has "id" (user) and "label" (long).
    # We want to group all users that share the same label, 
    # then produce a list of user IDs in each community.
    #
    # Steps:
    #   - convert the DF to an RDD
    #   - map each row => (label, id)
    #   - groupByKey => label => all users with that label
    #   - sort each community's users lexicographically
    #   - collect all communities as a list of user-lists
    #   - sort them by the rules in the PDF:
    #        1) ascending length
    #        2) if two communities have the same length, compare by
    #           the lexicographically smallest user in that community.
    #   - write them out
    # 
    # CAREFUL with multiple sorts. We'll do a single sort with a tuple:
    #   (len(community), community[0])
    # where community[0] is the first user in lexicographic order.
    lpa_rdd = lpa_result_df.rdd.map(lambda row: (row["label"], row["id"]))
    # groupByKey => label => [user1, user2, ...]
    label_groups = lpa_rdd.groupByKey().mapValues(list)  # => (label, [users...])
    
    # For each (label, [users...]), sort the user IDs, then transform to just that sorted list
    # We'll ignore the label after we produce the sorted user list.
    communities_rdd = label_groups.map(lambda x: sorted(x[1]))

    # Next, we want to sort these communities by:
    #   1) ascending length
    #   2) ascending first user 
    # Because the list is already sorted inside, the first user is at index [0].
    # So the final sort key is (len(community), community[0]).
    sorted_communities = communities_rdd.sortBy(lambda community: (len(community), community[0]))

    # -------------------------------------
    # 9. Write the output file
    # -------------------------------------
    # The PDF says each community line should be of the form:
    #   'uid1', 'uid2', 'uid3', ...
    # So we will format them carefully with quotes and commas,
    # ensuring that within a community, the user IDs are also lex sorted 
    # (which we did above).
    final_communities = sorted_communities.collect()
    
    with open(output_file_path, 'w') as out_file:
        for community in final_communities:
            # community is like ['userA', 'userB', 'userC'] 
            # We want the line to read: 'userA', 'userB', 'userC'
            # We'll do a join with "', '" as the separator, 
            # then add single quotes around the entire thing properly.
            line = ", ".join(f"'{u}'" for u in community)
            out_file.write(line + "\n")

    # -------------------------------------
    # 10. Print duration
    # -------------------------------------
    end_time = time.time()
    duration = end_time - start_time
    print("Duration:", duration)

    # Stop the SparkContext just to be clean.
    sc.stop()

# ------------------------------------------------------------------------------
# Standard Python convention: if we're running this file directly,
# call main(). If we import it as a module, it won't run immediately.
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    main()
